### PR TITLE
MM-22067: Check guest email allowance separately

### PR DESCRIPTION
**Pull request by ethervoid**
Right now the guest creation is checking the email allowance
using the team settings but it should be done using the guest settings
in this case

#### Summary

This PR changes how the email allowance is checked for guest users.

Right now is using the specific team + signup settings but in the guest case should be used the specific guest whitelist rules.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-22067

### CODE REVIEW

File: app/team.go [hunk 1]
```diff
@@ -88,11 +88,28 @@ func (a *App) isTeamEmailAddressAllowed(email string, allowedDomains string) boo
 	return true
 }
 
+func (a *App) isGuestEmailAddressAllowed(email string) bool {
```
Comment (agnivade) [inline app/team.go]: This new method has a lot of overlap with the old `isTeamEmailAddressAllowed` method. Why don't we rename the old method to `isEmailAddressAllowed` and pass `*model.User` to it, and depending on whether the user is guest or not, perform the appropriate logic. That way, there is less duplication and in future, one doesn't have to remember that there are 2 different functions depending on whether a user is guest or not.

File: app/team.go [hunk 2]
```diff
@@ -88,11 +88,28 @@ func (a *App) isTeamEmailAddressAllowed(email string, allowedDomains string) boo
...
+		}
+	}
+	return false
+}
+
 func (a *App) isTeamEmailAllowed(user *model.User, team *model.Team) bool {
 	if user.IsBot {
 		return true
 	}
 	email := strings.ToLower(user.Email)
```
Comment (agnivade) [inline app/team.go]: On an unrelated note, this is redundant. `ToLower` is already done in `isTeamEmailAddressAllowed`.

File: app/team_test.go [hunk 3]
```diff
@@ -269,6 +270,68 @@ func TestAddUserToTeamByToken(t *testing.T) {
...
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "channels": th.BasicChannel.Id}),
+		)
+		guestEmail := rguest.Email
+		rguest.Email = "test@restricted.com"
+		_, err := th.App.Srv.Store.User().Update(rguest, false)
+		require.Nil(t, err)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		_, err = th.App.AddUserToTeamByToken(rguest.Id, token.Token)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
```
Comment (agnivade) [inline app/team_test.go]: I don't believe we usually add a format message for an error. Perhaps we can remove the message ?

File: app/team_test.go [hunk 4]
```diff
@@ -269,6 +270,68 @@ func TestAddUserToTeamByToken(t *testing.T) {
...
+		th.App.UpdateConfig(func(cfg *model.Config) { *cfg.TeamSettings.RestrictCreationToDomains = "restricted.com" })
+		token := model.NewToken(
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "channels": th.BasicChannel.Id}),
+		)
+		_, err = th.App.Srv.Store.User().Update(rguest, false)
+		require.Nil(t, err)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		_, err = th.App.AddUserToTeamByToken(rguest.Id, token.Token)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
```
Comment (agnivade) [inline app/team_test.go]: Same as above.

File: app/team_test.go [hunk 5]
```diff
@@ -269,6 +270,68 @@ func TestAddUserToTeamByToken(t *testing.T) {
...
+		}()
+		th.App.UpdateConfig(func(cfg *model.Config) { *cfg.GuestAccountsSettings.RestrictCreationToDomains = "restricted.com" })
+		token := model.NewToken(
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "channels": th.BasicChannel.Id}),
+		)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		_, err := th.App.AddUserToTeamByToken(rguest.Id, token.Token)
+		require.NotNil(t, err)
+		errRegexp := regexp.MustCompile(`Email must be from a specific domain`)
```
Comment (agnivade) [inline app/team_test.go]: The error is an AppError. Let's check the id rather than do a regex match which is a bit brittle IMO. And also change the check to assert.Equal from require.Equal.

File: app/user_test.go [hunk 6]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, forbiddenDomainToken)
+		require.NotNil(t, err)
+		require.Nil(t, newGuest)
+		errRegexp := regexp.MustCompile(`does not belong to an accepted domain`)
+		require.Equal(t, "does not belong to an accepted domain", errRegexp.FindString(err.Error()))
```
Comment (agnivade) [inline app/user_test.go]: Same as above. Let's check the error id, rather than a string match.

File: app/user_test.go [hunk 7]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+
+		newGuest, err = th.App.CreateUserWithToken(&guest, grantedDomainToken)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
```
Comment (agnivade) [inline app/user_test.go]: Same as above. Do we need the format message ?

File: app/user_test.go [hunk 8]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, grantedInvitationEmail, newGuest.Email, "The user email must be the invitation one")
```
Comment (agnivade) [inline app/user_test.go]: "The user email must be the same as the one used in the invitation."

File: app/user_test.go [hunk 9]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+		_, err = th.App.Srv.Store.Token().GetByToken(grantedDomainToken.Token)
+		require.NotNil(t, err, "The token must be deleted after be used")
+
+		members, err := th.App.GetChannelMembersForUser(th.BasicTeam.Id, newGuest.Id)
```
Comment (agnivade) [inline app/user_test.go]: Any reason we are checking this ? Seems unrelated. I do note we are doing this in other tests too. But I would like to understand why, and if there isn't a need, then not do it.

File: app/user_test.go [hunk 10]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, token)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, invitationEmail, newGuest.Email, "The user email must be the invitation one")
+		_, err = th.App.Srv.Store.Token().GetByToken(token.Token)
+		require.NotNil(t, err, "The token must be deleted after be used")
```
Comment (agnivade) [inline app/user_test.go]: after it's used / after it has been used.

File: app/user_test.go [hunk 11]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+			Email:       strings.ToLower(model.NewId()) + "+test@example.com",
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, token)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, invitationEmail, newGuest.Email, "The user email must be the invitation one")
```
Comment (agnivade) [inline app/user_test.go]: Same message as above. Or we can just remove it. The user will see the line number anyways. Up to you.

File: app/user_test.go [hunk 12]
```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
...
+		_, err = th.App.Srv.Store.Token().GetByToken(token.Token)
+		require.NotNil(t, err, "The token must be deleted after be used")
+
+		members, err := th.App.GetChannelMembersForUser(th.BasicTeam.Id, newGuest.Id)
```
Comment (agnivade) [inline app/user_test.go]: Same concern as above.

File: app/team.go (hunk 1 above)
Comment (ethervoid) [reply app/team.go]: Yes I know. I though about the two approaches and finally I've taken mainly for two reasons:

- Having all the logic in the same method looks a bit confusing. We have to create a variable depending on the user.isGuest, getting different configuration and so on.
- Even though right now the code is duplicated we can leverage that in the future is that changes. A bit of duplication is not hard

But I've found a better way to do a mix of the proposed solutions :)

File: app/team.go (hunk 2 above)
Comment (ethervoid) [reply app/team.go]: Good catch

File: app/team_test.go (hunk 3 above)
Comment (ethervoid) [reply app/team_test.go]: Ok, I've been following the old requires

File: app/team_test.go (hunk 5 above)
Comment (ethervoid) [reply app/team_test.go]: Oh, didn't know about the err.Id. Much better indeed

File: app/user_test.go (hunk 7 above)
Comment (ethervoid) [reply app/user_test.go]: Removed

File: app/user_test.go (hunk 9 above)
Comment (ethervoid) [reply app/user_test.go]: Basically is checking if the uses was added to the team as a form of verification

File: app/user_test.go (hunk 8 above)
Comment (ethervoid) [reply app/user_test.go]: Removed

File: app/team.go (hunk 1 above)
Comment (ethervoid) [reply app/team.go]: Yes. I though a bit about how to handle it.

I dismissed to include everything into the same method because it seems to broke the single responsibility a bit too much: allowed domains logic, email normalization logic, domain belonging logic...

But you're right, my decision wasn't better so I've came up with a much better approach I hope you like it as well :)

File: app/team.go [hunk 13]
```diff
@@ -93,7 +91,17 @@ func (a *App) isTeamEmailAllowed(user *model.User, team *model.Team) bool {
...
+	return a.isEmailAddressAllowed(email, allowedDomains)
+}
+
+func (a *App) getAllowedDomains(user *model.User, team *model.Team) []string {
+	if user.IsGuest() {
+		return []string{*a.Config().GuestAccountsSettings.RestrictCreationToDomains}
+	} else {
+		// First check per team allowedDomains, then app wide restrictions
+		return []string{team.AllowedDomains, *a.Config().TeamSettings.RestrictCreationToDomains}
+	}
```
Comment (reflog) [inline app/team.go]: wouldn't it look cleaner as:
```go
result := []string{*a.Config().GuestAccountsSettings.RestrictCreationToDomains}
if !user.IsGuest() {
		// First check per team allowedDomains, then app wide restrictions
		result = append([]string{team.AllowedDomains}, result...)
	}
return result
```

Comment (ethervoid) [reply app/team.go]: I tend to avoid re-assigning variable state but maybe this way is better or idiomatic for go. What do you think @agnivade ?

Comment (agnivade) [reply app/team.go]: Idiomatic Go is to just outdent the else condition:

```go
if user.IsGuest() {
	return []string{*a.Config().GuestAccountsSettings.RestrictCreationToDomains}
} 
// First check per team allowedDomains, then app wide restrictions
return []string{team.AllowedDomains, *a.Config().TeamSettings.RestrictCreationToDomains}
```

File: app/team.go (hunk 1 above)
Comment (agnivade) [reply app/team.go]: Looks great now.

File: app/team_test.go [hunk 14]
```diff
@@ -269,6 +269,67 @@ func TestAddUserToTeamByToken(t *testing.T) {
...
+		}()
+		th.App.UpdateConfig(func(cfg *model.Config) { *cfg.GuestAccountsSettings.RestrictCreationToDomains = "restricted.com" })
+		token := model.NewToken(
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "channels": th.BasicChannel.Id}),
+		)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		_, err := th.App.AddUserToTeamByToken(rguest.Id, token.Token)
+		require.NotNil(t, err)
+		require.Equal(t, "api.team.join_user_to_team.allowed_domains.app_error", err.Id)
```
Comment (agnivade) [inline app/team_test.go]: I think this can be "assert.Equal". This does not block the test from moving ahead. Just an assertion failure.

File: app/team_test.go [hunk 15]
```diff
@@ -269,6 +269,67 @@ func TestAddUserToTeamByToken(t *testing.T) {
...
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "channels": th.BasicChannel.Id}),
+		)
+		guestEmail := rguest.Email
+		rguest.Email = "test@restricted.com"
+		_, err := th.App.Srv.Store.User().Update(rguest, false)
+		require.Nil(t, err)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		_, err = th.App.AddUserToTeamByToken(rguest.Id, token.Token)
+		require.Nil(t, err, "Should add user to the team")
```
Comment (agnivade) [inline app/team_test.go]: Did you mean to keep the message ? It's not a big deal anyways, just asking.

File: app/user_test.go [hunk 16]
```diff
@@ -653,6 +653,88 @@ func TestCreateUserWithToken(t *testing.T) {
...
+			Email:       strings.ToLower(model.NewId()) + "+test@example.com",
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, forbiddenDomainToken)
+		require.NotNil(t, err)
+		require.Nil(t, newGuest)
+		require.Equal(t, "api.user.create_user.accepted_domain.app_error", err.Id)
```
Comment (agnivade) [inline app/user_test.go]: Same as above. This can be assert.Equal. If this fails, the entire test will be stopped, which is not what we want I believe.

File: app/user_test.go [hunk 17]
```diff
@@ -653,6 +653,88 @@ func TestCreateUserWithToken(t *testing.T) {
...
+		require.Nil(t, err, "Should add user to the team")
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, grantedInvitationEmail, newGuest.Email)
+		_, err = th.App.Srv.Store.Token().GetByToken(grantedDomainToken.Token)
+		require.NotNil(t, err)
+
+		members, err := th.App.GetChannelMembersForUser(th.BasicTeam.Id, newGuest.Id)
+		require.Nil(t, err)
+		require.Len(t, *members, 1)
+		require.Equal(t, (*members)[0].ChannelId, th.BasicChannel.Id)
```
Comment (agnivade) [inline app/user_test.go]: assert.Equal

File: app/user_test.go [hunk 18]
```diff
@@ -653,6 +653,88 @@ func TestCreateUserWithToken(t *testing.T) {
...
+			Email:       strings.ToLower(model.NewId()) + "+test@example.com",
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, token)
+		require.Nil(t, err, "Should add user to the team")
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, invitationEmail, newGuest.Email, "The user email must be the invitation one")
```
Comment (agnivade) [inline app/user_test.go]: assert.Equal

File: app/user_test.go [hunk 19]
```diff
@@ -653,6 +653,88 @@ func TestCreateUserWithToken(t *testing.T) {
...
+		_, err = th.App.Srv.Store.Token().GetByToken(token.Token)
+		require.NotNil(t, err)
+
+		members, err := th.App.GetChannelMembersForUser(th.BasicTeam.Id, newGuest.Id)
+		require.Nil(t, err)
+		require.Len(t, *members, 1)
+		require.Equal(t, (*members)[0].ChannelId, th.BasicChannel.Id)
```
Comment (agnivade) [inline app/user_test.go]: assert.Equal.
